tdt-driver/man/xbcomini.1
.TH XBCOMINI 1
.SH NAME
xbcomini \- set communication baud rate for Tucker Davis XBUS devices
.SH SYNOPSIS
.B xbcomini 
[
.I ioport
]
[
.I baudrate
]
.SH DESCRIPTION
.PP
.I xbcomini 
initializes the communications baud rate and protocol
of both the Tucker Davis XBUS devices and the port on your computer 
through which you will communicate with the XBUS.  

.I xbcomini 
must be run each time you turn on your computer.

Perform the following steps, in order, to successfully use 
.I xbcomini:
.TP
.B 1.
Push the "ID" button on each XB1 rack for three seconds until the amber
ACTIVITY LED stops blinking and remains lit.
.TP
.B 2.
Run
.I xbcomini.  
The ACTIVITY LED on each XB1 rack will go off, and all 
devices will be reset to the specified baud rate.

.SH OPTIONS
.I ioport
.RS
This is the port through which your computer 
communicates with the XBUS,  and must be one of "com1", "com2", com3", "com4",
"apa" or "apb".

"com1" through "com4" refer to the physical ports COM1 through COM4.  Use "apa" if you
have an AP2 Array Processor board installed with jumper JP1 set to 
option A (board uses ports 220-238H), and "apb" if JP1 was set to 
option B (board uses ports 240-250H).

If this parameter is not supplied, 
.I xbcomini 
reads the environment variable XBCOM, if set
to determine how to communicate with the XBUS.

.RE

.I baudrate
.RS
If this parameter is not supplied, 
.I xbcomini 
will initialize the baud rate to 38400 baud, and the protocol to 
1 start bit, 8 data bits, 1 stop bit.  

If the 
.I baudrate
parameter is supplied, it must be one of 1200, 2400, 4800, 9600, 19200
or 38400.
.RE
.SH FILES
None.
.SH ENVIRONMENT
.B XBCOM
.RS
This can be set to the communcation channel to use to communicate with 
the XBUS. It must be one of "com1", "com2", com3", "com4",
"apa" or "apb".
.SH DIAGNOSTICS
None.
.SH BUGS
None. (Yeah, sure.)
.SH AUTHOR
Tucker Davis Technologies, Inc.  <quikki@tdt-quikki.com> 
Adapted to Linux by Chris Malek <cmalek@etho.caltech.edu>
.SH "SEE ALSO"
.BR xbsysid (1),
.BR xbidchck (1),
.BR apld (1),




iserver/Changes
1.2.1:
	tdtproc now checks the environment variable DAQ_AP2_PORT to determine
	which port to use for apinit(); defaults to 'APa'
1.2.2:
	forgot to do the above for XB1init().  Now it does this.
1.2.3:
	made is-matlab.c delete old is_info array before inserting new one
	fixed " characters in set_att.m
1.2.4:
	fixed bug in is-matlab.c:load(): was only copying part of the input
	  waveform to the DA buffer.
	changed is_matlab.c:get_field_int() to check if the field exists
	  before trying to read it, so is_init will not segfault now when
	  you pre-initialize some fields (e.g. is_init.debugflag)
1.2.5:
	glibc > 2.0 does something screwy with stat(), so the call to stat()
	in is-tdt.c:file_is_executable() has been #ifdef'd out
1.2.6:
	in tdtproc.c, removed some extremely uneccesary calls from arm_equipment()
	which were simply chewing up time: program_da1() and check_da_buflen().

	also in tdtproc.c:arm_equipment(), fixed case in ad arming switch() 
1.2.7:
	changed att() in is-matlab: now only takes 1x2 array as its 
	argument, instead of 'left' and 'right' type settings.
1.2.8:
	documented the is_*.m files better, fixed is_record.m
	fixed an error message in is-tdt.c
1.2.9:
	iserver now uses GNU libtool to compile as a shared library on
	both SunOS and Linux.
1.2.10:
	added LED code, removed sync_pulse dependency on ET1
1.2.11:
	Moved matlab install directory to /usr/local/local-matlab

	Release 2:
	Recompiled mex files for Matlab 6
1.2.12:
	Fixed tdtproc/iserver_test.c so that it actually works.
	iserver_test is now part of the distributed package.
	Added code to tdtproc to write DA bufs to files, both before
		and after writing to the AP2 play buffers (DAQ_DA_DUMP and 
		DAQ_AP2_DA_DUMP)


tdt-driver/utils/xbcomini.c
#include <stdlib.h>
#include <math.h>
#ifndef __linux__
#include <bios.h>
#endif /* __linux__ */
#include <stdio.h>
#include <dos.h>
#include <string.h>
#include "xbdrv.h"

#define TXPORT			0
#define RXPORT			0
#define BAUDLOW			0
#define BAUDHIGH		1
#define INTENABLE	1
#define INTID			2
#define LINECONT		3
#define MODEMCONT	4
#define LINESTAT		5
#define MODEMSTAT		6

#define B300 384
#define B1200 96
#define B2400 48
#define B4800 24
#define B9600 12
#define B19200 6
#define B38400 3

#define upper(x) ((x & 0xff00)>>8)
#define lower(x) (x & 0xff)

void usageerr(void)
{
	printf("\n\nUsage:  XBCOMINI <ioport> <baudrate>\n\n");
	printf("    ioport => APa, APb, COM1, COM2, COM3 or COM4.\n");
	printf("  baudrate => 1200, 2400, 4800, 9600, 19200 or 38400.\n\n");
	exit(0);
}

int linecontrolport, baudhiport, baudloport, linestatusport;
int intenableport, modemcontrolport;

#ifndef __linux__
/* --------------------------------------------------------------------- */
/* Linux note: This section must be adapted to use the /dev/cua[0-3]     */
/*				devices; see setserial(8) code for ideas on how to       */
/*				set the baud rate for the UART of the COM port           */
/*              You could also use tcsetattr(2).                         */
/* --------------------------------------------------------------------- */
/*************************************************
**		comset(baud,comport)			  **
**		does initialization of specified comport  **
**		parity is fixed to NONE, word size is	  **
**		8 bits, with 1 stop bit.		  **
**						**
**		parameters:				  **
**		baud -- any of the following baud rate	  **
**		values are legal.  1200,2400,4800,	  **
**		9600,19200,38400.			  **
**		comport -- 1 - 4 are recognized for COM1  **
**		through COM4				  **
*************************************************/
void comset(long baud)
{
	char commandbyte;
	int baudval;
	unsigned char temp;


	if ((baud != 1200) && (baud != 2400) &&
		(baud != 4800) && (baud != 9600) && (baud != 19200) && (baud != 38400))
	{
		printf("\n\n\n	Illegal Baud Rate Specified.");
		usageerr();
	}

	if ((cc < 1) || (cc > 4))
	{
		printf("\n\n\n	Illegal COM Port Specified.");
		usageerr();
	}

	switch(cc)
	{
		case 1:
					cpa[0] = COM1base;
					break;
		case 2:
					cpa[0] = COM2base;
					break;
		case 3:
					cpa[0] = COM3base;
					break;
		case 4:
					cpa[0] = COM4base;
					break;
	}

	switch(baud)
	{
		case 1200:
					baudval = B1200;
					break;

		case 2400:
					baudval = B2400;
					break;

		case 4800:
					baudval = B4800;
					break;

		case 9600:
					baudval = B9600;
					break;

		case 19200:
					baudval = B19200;
					break;

		case 38400:
					baudval = B38400;
					break;
		default:
					usageerr();
	}

	linecontrolport = cpa[0] + LINECONT;
	baudhiport = cpa[0] + BAUDHIGH;
	baudloport = cpa[0] + BAUDLOW;
	linestatusport = cpa[0] + LINESTAT;
	intenableport = cpa[0] + INTENABLE;
	modemcontrolport = cpa[0] + MODEMCONT;

	/* Linux note: bioscom(cmd, byte, port) sets no parity, 9 bits, 1 
	   stop bit and 9600 baud rate of the UART for the apropriate COM port, 
	   I think */
	commandbyte = 0xE3;	/* n-8-1 9600 */
	bioscom(0,commandbyte,cc-1);

	outportb(intenableport,0x00);
	temp = inportb(linecontrolport);
	temp = temp | 0x80;
	outportb(linecontrolport,temp);
	outportb(baudloport, lower(baudval));
	outportb(baudhiport, upper(baudval));
	temp = temp & 0x7f;
	outportb(linecontrolport,temp);
}
#endif /* __linux__ */



void apcomset(void)
{
	switch(cc)
	{
		case COM_APa:
		{
			cc = COM_APa;
			cpa[0] = 0x230;
			break;
		}
		case COM_APb:
		{
			cc = COM_APb;
			cpa[0] = 0x250;
			break;
		}
		default:
		{
			printf("\n\n\n	Illegal COM Port Specified.");
			printf("\n	Use:  COM1, COM2, COM3, COM4, APa or APb.");
			printf("\n\n\n");
			exit(0);
		}
	}
	io_init(cpa[0]);
	outportb(cpa[0],3);
	outportb(cpa[0],3);
	outportb(cpa[0],3);
	outportb(cpa[0],0x15);
	outportb(cpa[0],0x15);
	outportb(cpa[0],0x15);
}


void main(int argc, char *argv[], char *env[])
{
	 char *s;
	 long baudrate;
	 int i,j;
	 int xx;

	 xbtimeout = (long)XB_TIMEOUT * 100;

	 cc = COM_APa;

	 s = getenv("XBCOM");
	 if (s != NULL) {
#ifndef __linux__
		 if (!strncmp(s,"COM1", 4) || !strncmp(s,"com1"), 4)
			 cc = COM1;
		 if (!strncmp(s,"COM2", 4) || !strncmp(s,"com2"), 4)
			 cc = COM2;
		 if (!strncmp(s,"COM3", 4) || !strncmp(s,"com3"), 4)
			 cc = COM3;
		 if (!strncmp(s,"COM4", 4) || !strncmp(s,"com4"), 4)
			 cc = COM4;
#endif /* __linux__ */
		 if (!strncmp(s,"APb", 3) || !strncmp(s,"apb", 3))
			 cc = COM_APb;
	 }

	 baudrate = 38400;
	 xx = 0;

	 if (argc > 1) {
		 if (!strncmp(argv[1],"APa",3) || !strncmp(argv[1],"apa",3) || 
			 !strncmp(argv[1],"APA",3))
			 xx = COM_APa;
		 if (!strncmp(argv[1],"APb",3) || !strncmp(argv[1],"apb",3) || 
			 !strncmp(argv[1],"APB",3))
			 xx = COM_APb;
#ifndef __linux__
		 if (!strncmp(argv[1],"COM1",4) || !strncmp(argv[1],"com1",4))
			 xx = COM1;
		 if (!strncmp(argv[1],"COM2",4) || !strncmp(argv[1],"com2",4))
			 xx = COM2;
		 if (!strncmp(argv[1],"COM3",4) || !strncmp(argv[1],"com3",4))
			 xx = COM3;
		 if (!strncmp(argv[1],"COM4",4) || !strncmp(argv[1],"com4",4))
			 xx = COM4;
#endif /* __linux__ */

		 if (!xx)
			 sscanf(argv[1],"%ld",&baudrate);
		 else
			 cc = xx;
	 }

	 if (argc==3)
	 {
		 if (!xx)
		 {
			 printf("\n\nIllegal argument specified.");
			 usageerr();
		 }
		 sscanf(argv[2],"%ld",&baudrate);
	 }

#ifndef __linux__
	 if (cc<COM_APa)
		 comset(baudrate);
	 else
#endif /*__linux__*/
		 apcomset();

	 for(i=0; i<100; i++)
	 {
		 iosend(BAUD_LOCK);
		 for(j=0; j<10000; j++);
	 }
}

